Add `notify-on-failure` command

masutaka opened the pull request:
Hi @pbrisbin,

I want to notify me when the CI of the master branch fails. So, I added `notify-on-failure` command.

I really wanted to add `notify-on-failure` **job**, but I gave up because CircleCI is impossible to run a final job in workflow unconditionally.
See also https://discuss.circleci.com/t/how-to-always-run-a-sequential-final-job-e-g-slack-notification/35362

I referred to Slack orb:

* https://github.com/CircleCI-Public/slack-orb/blob/v3.4.2/src/commands/notify-on-failure.yml
    * https://github.com/CircleCI-Public/slack-orb/blob/v3.4.2/src/commands/status.yml

Thanks for the useful orb!

masutaka committed:
Add parameters `fail_only` and `only_for_branches` to `notify` command

masutaka committed:
Fix stile

closes #2

pbrisbin commented on src/commands/notify.yml:
```diff
@@ -24,18 +24,78 @@ parameters:
+    description: If `true`, notifications successful jobs will not be sent
+    type: boolean
+    default: false
+  only_for_branches:
```
Is there any benefit of this over using the built-in branch filtering?

For example:

```yaml
      - pushover/notify:
          name: notify
          requires: [release-frontend, release-backend]
          title: 'Restyled deployments'
          message: 'restyled.io deployed at ${CIRCLE_SHA1:0:10}'
          filters:  # <-- here
            branches:
              only:
                - master
```

masutaka replied:
I want to be notified when a job I created fails. So it needs to be implemented in the command, not the job.

This is a use case:
https://github.com/masutaka/circleci-bundle-update-pr/blob/fd3401ec320a6b72de9a83a11f69d6d499318a5c/.circleci/config.yml

FYI: It can't specify `filters` in the CircleCI command.
https://circleci.com/docs/2.0/configuration-reference/#commands-requires-version-21

pbrisbin commented on src/commands/notify.yml:
```diff
@@ -24,18 +24,78 @@ parameters:
+      when: on_success
+
+  - run:
+      name: Provide error if non-bash shell
```
I think I'd probably leave out this complexity. If `{the thing set as shell}` isn't installed, I'd expect the Circle user-experience to be good enough that we don't have to provide our own check and error message.

masutaka replied:
I thought a little bit of that too, but this time I followed the slack-orb. I can remove it.
https://github.com/CircleCI-Public/slack-orb/blob/v3.4.2/src/commands/status.yml#L79-L85

pbrisbin commented on src/commands/notify.yml:
```diff
@@ -24,18 +24,78 @@ parameters:
+      name: Pushover - Setting Success Condition
+      command: |
+        echo 'export PUSHOVER_BUILD_STATUS="success"' >> $BASH_ENV
+      when: on_success
```
Is there a reason we can't ask users who want this behavior to do it themselves?

```yaml
      - pushover/notify:
          name: notify
          requires: [...]
          title: '...'
          message: 'Build failed'
          when: on_failure
```

We may need to accept our own `when` parameter and say `when: <<parameters.when>>`?

If none of that works, I wonder if making a new `pushover/notify-on-failure` command for this use-case would be a simpler option.

masutaka replied:
I want to be notified only when the following `run(s)` have failed by line 87.
https://github.com/masutaka/circleci-bundle-update-pr/blob/fd3401ec320a6b72de9a83a11f69d6d499318a5c/.circleci/config.yml#L73-L88

pbrisbin reviewed (changes requested):
Thanks very much for this, it's indeed a great use-case for this orb. I _think_ we might be able to achieve the same functionality in a simpler way -- or even by doing nothing at all, except maybe adding some example docs. Let me know what you think!

pbrisbin commented:
I just clicked through the slack-orb links, sorry I missed them the first time. There's probably a reason they do it that way as well, so I may be missing something in my earlier comments.

I'm happy to accept this change, but could you:

1. Leave the current `pushover/notify` command alone, and add a dedicated `notify-on-failure` command (just like the Slack one)
1. Have the `notify-on-failure` command define default messages for you, with details about the Job (also just like the Slack one), and then call `notify` to accomplish the actual notification

I think this would be a nice way to re-use the simpler thing we have, leave it there for lower-level use-cases, and build a higher-level / convenient command on top.

masutaka commented:
The idea of making a new `pushover/notify-on-failure` command looks good. I will implement and add commit(s).

Thanks!

masutaka committed:
Revert "Add parameters `fail_only` and `only_for_branches` to `notify` command"

This reverts commit 72817ce5f2e2b6de87f09795915ee8b9dffe2ed8.

masutaka committed:
Add command `notify-on-failure`

masutaka committed:
Add example `notify-on-failure`

masutaka commented on src/commands/notify-on-failure.yml:
```diff
@@ -0,0 +1,78 @@
+      when: on_success
+      name: Pushover - Setting Success Condition
+      command: |
+        echo 'export PUSHOVER_BUILD_STATUS="success"' > /tmp/PUSHOVER_JOB_STATUS
```
This is an idea of Slack orb. I can remove the dependency with /bin/bash (BASH_ENV).
https://github.com/CircleCI-Public/slack-orb/blob/v4.2.0/src/commands/notify.yml#L61

masutaka commented on src/commands/notify-on-failure.yml:
```diff
@@ -0,0 +1,78 @@
+  message:
+    description: Notification message
+    type: string
+    default: "Failed ${CIRCLE_USERNAME:=unknown}'s build (#${CIRCLE_BUILD_NUM}) in ${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME} (${CIRCLE_BRANCH})"
```
`CIRCLE_USERNAME` is sometimes empty.
e.g. scheduled job

masutaka commented:
@pbrisbin

Sorry for the too delay. :bow:

I have implemented `notify-on-failure` command instead of modifying `notfiy` command. It has the default `title` and `message` for failure. And the `priority` is 1 (high).

Also, I added `only_for_branches` because it is impossible to set `filters` to a job like I can to [workflow](https://circleci.com/docs/2.0/configuration-reference/#workflows).

What do you think?

FYI:

The current Slack orb (v4.2.0) has only `notify` command.
https://github.com/CircleCI-Public/slack-orb/tree/v4.2.0/src/commands

Previously, Slack orb (v3.4.2) had commands such as `notify-on-failure` as well as `notify`.
https://github.com/CircleCI-Public/slack-orb/tree/v3.4.2/src/commands

pbrisbin commented:
Thanks!

By the way, I'm mostly moving things to GitHub Actions, so I'm not going to be actively working on this Orb. Would you like to be added as a maintainer? I'm also happy to let you take to code to your own repo, and set up your own Orb -- whatever works best for you.

masutaka commented:
Welcome!

I see. So from now on, I'll fork this repository and publish `masutaka/pushover` orb .

Thank you for the useful orb!

pbrisbin commented:
> from now on, I'll fork this repository and publish masutaka/pushover orb .

:+1: Please let me know when the Orb is available, and I will Archive this repository with a link.

masutaka commented:
Just released v2.0.0 :tada:

* https://circleci.com/developer/orbs/orb/masutaka/pushover
* https://github.com/masutaka/pushover-orb
